packages/runtime/container-runtime/src/batchManager.ts

```diff
@@ -84,4 +89,21 @@ export class BatchManager {
             },
         };
     }
+
+    private checkReferenceSequenceNumber(message: BatchMessage) {
+        if (this.mc.config.getBoolean("Fluid.ContainerRuntime.DisableOpReentryCheck") === true) {
+            return;
+        }
+
+        if (this.pendingBatch.length > 0
+            && message.referenceSequenceNumber !== this.pendingBatch[0].referenceSequenceNumber) {
+            throw new GenericError(
```

Comment:
Will this prohibit DDS code itself from submitting ops during processing? We actually have several instances of this in SharedTree (and we are careful to avoid any inconsistencies).

Reply:
I also leverage this in a DDS from an FHL project that I'll be spinning up a PR for soon, so I'd love to chat about this.

Reply:
Hi @taylorsw04 , can you point me to an example of this in SharedTree? Are the scenarios covered by test and are they in the same monorepo as the client code in this PR? If tests pass then I don't believe this should block your scenarios.

The gist of it is that this validation ensures that all ops in the batch have been created while having the same view of the world (the reference sequence number). On a case-by-case basis, it might be safe or it might not be, hence it should be regarded as undefined behavior and the proposal here is to block it when the op is produced.

> we are careful to avoid any inconsistencies

I'm interested in how this is accomplished. Can you point me to the relevant logic?

FYI @vladsud

Reply:
I believe they happen to pass because they spin up async work (blob uploading) before sending that off in an op. This is true in all cases that I know of, FWIW, but I don't think that's worth focusing on as I think there are viable reasons that a DDS can shoulder the burden of handling this case (although you might have ideas on how to accomplish the same, but safer).

An example: https://github.com/microsoft/FluidFramework/blob/main/experimental/dds/tree/src/SharedTree.ts#L1280

Reply:
@taylorsw04, just to be sure - are you saying it's Ok to push that requirement to DDSs not to issue ops while processing ops? Or reverse?

Reply:
If re-entrancy is allowed, then even if DDS is very careful here and does "all the right things", an event callback causing re-entrancy results in callbacks running in reverse order for _some_ users of the system. This is very dangerous for correctness.

I.e. image situation that there are 2 callbacks registered - A & B.
If on change # 1, callback A runs first and makes another change # 2, then we will see this flow:

A for # 1
A for # 2
B for # 2
B for # 1

B is getting events in wrong order.

Reply:
@vladsud I am saying the reverse. I do think your scenario is dangerous, and should probably never be implemented. However, in the case of our DDSs, we are submitting ops that are only mutating internal state and never invoke callbacks (both are related to persisting state via uploading blobs). So, I agree that event-calling re-entrancy is likely _always_ bad, but we aren't in that bucket.

Reply:
The behavior is now opt-in, so by default this is allowed. So at least for now, the default will allow sending ops from event handlers. Even if explicitly disallowed, there's a feature flag which can override it at runtime. So, in any case, your scenario will not be blocked, regardless of how we go forward (as you'd always have an option to turn it off, either from code or from feature flags - preferably the former).

However, I'm interested in your scenario and how would it differ from the example Vlad described. When you say internal state, do you mean the state of another DDS or the current DDS? Or state which doesn't fire another event handler?

Reply:
We could make change design and work off events only. I.e. allow any op sending, but disallow re-entrancy in events across container runtime. But I think that will cut off scenario that might be useful - having an "uber-DDS" (or "channel") that is an aggregation of multiple child DDSs. In such system, lower level changes (and events) are translated to higher level events on the parent, and it's Ok to have event reentrancy.

So, it feels like catching these issues at the op submission level is a bit better approach. And so far, it sounds like we have not seen an example where pushing people to do async op submission is a problem.

Reply:
@andre4i a good example of this was the flushing to storage that the partial map DDS does in response to receiving an op (on some cadence). The key here is that it will never invoke event handlers and is only persisting a stable view of the internals of the DDS (no reliance on other DDS' states).

Regardless, I'm happy with the changes.